plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs

```diff
@@ -42,20 +58,75 @@ import Data.Set qualified as Set
 import Data.Tuple (swap)
 import GHC.Generics (Generic)
 import Ledger (OnChainTx (..), SlotRange, SomeCardanoApiTx, Tx (..), txId)
+import Ledger.Tx.CardanoAPI (fromCardanoAddress, fromCardanoTxOutDatum, fromCardanoValue, toCardanoTxOutBabbage,
+                             toCardanoTxOutDatumHashBabbage)
 import Plutus.Script.Utils.V1.Scripts (datumHash, mintingPolicyHash, redeemerHash, validatorHash)
-import Plutus.V1.Ledger.Api (Address, Datum, DatumHash, MintingPolicy (getMintingPolicy),
-                             MintingPolicyHash (MintingPolicyHash), Redeemer, RedeemerHash, Script, TxId,
-                             TxOut (txOutAddress), TxOutRef (TxOutRef), Validator (getValidator),
+import Plutus.V1.Ledger.Api (Datum, DatumHash, MintingPolicy (getMintingPolicy), MintingPolicyHash (MintingPolicyHash),
+                             Redeemer, RedeemerHash, Script, TxId, TxOutRef (TxOutRef), Validator (getValidator),
                              ValidatorHash (ValidatorHash))
 import Plutus.V1.Ledger.Scripts (ScriptHash (ScriptHash))
 import Plutus.V1.Ledger.Tx (TxIn (txInType), TxInType (ConsumeScriptAddress))
+import Plutus.V2.Ledger.Api (Address, OutputDatum, Value)
 import Prettyprinter
 
+data ReferenceScript = ReferenceScriptNone | ReferenceScriptInAnyLang C.ScriptInAnyLang
+  deriving (Eq, Show)
+
+instance ToJSON ReferenceScript where
+  toJSON (ReferenceScriptInAnyLang s) = object ["referenceScript" .= s]
+  toJSON ReferenceScriptNone          = Aeson.Null
+
+instance FromJSON ReferenceScript where
+  parseJSON = Aeson.withObject "ReferenceScript" $ \o ->
+    case Aeson.lookup "referenceScript" o of
+      Nothing        -> pure ReferenceScriptNone
+      Just refScript -> ReferenceScriptInAnyLang <$> parseJSON refScript
+
+fromCardanoTxOutRefScript :: C.ReferenceScript era -> ReferenceScript
+fromCardanoTxOutRefScript = \case
+    C.ReferenceScriptNone      -> ReferenceScriptNone
+    C.ReferenceScript _ script -> ReferenceScriptInAnyLang script
+
+data ChainIndexTxOut = ChainIndexTxOut
+  { citoAddress   :: Address -- ^ We can't use AddressInAnyEra here because of missing FromJson instance for Byron era
+  , citoValue     :: Value
+  , citoDatum     :: OutputDatum
+  , citoRefScript :: ReferenceScript
```

Comment:
For these three, why not reuse the ones from `Cardano.Api`?

Reply:
They are very annoying as they are parametrized with the era.

We would need to convert `RefinementScript era` to `RefinementScript BabbageEra` anyway, so I decided to define chain index's ADT for easier code and better control.

Datum misses JSON instances for example. They are not exported.

If it's not worth it, I could try again to use the cardano-api's types.

Reply:
Not at all. Let's keep it this way for now.

Reply:
Another option could be `PlutusLedgerApi.V2.TxOut`?

Reply:
`TxOut` is designed for on-chain use. Thus, the reference script part of the tx output only contains the hash and not the actual script (which is what the spec requires of it).